src/Features/Core/Portable/SimplifyTypeNames/SimplifyTypeNamesDiagnosticAnalyzerBase.cs

```diff
@@ -183,5 +195,55 @@ internal static Diagnostic CreateDiagnostic(SemanticModel model, OptionSet optio
 
         public DiagnosticAnalyzerCategory GetAnalyzerCategory()
             => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        private class AnalyzerImpl
+        {
+            private readonly SimplifyTypeNamesDiagnosticAnalyzerBase<TLanguageKindEnum> _analyzer;
+            private readonly ConcurrentDictionary<SyntaxTree, (StrongBox<bool> completed, SimpleIntervalTree<TextSpan, TextSpanIntervalIntrospector>? intervalTree)> _codeBlockIntervals = new ConcurrentDictionary<SyntaxTree, (StrongBox<bool> completed, SimpleIntervalTree<TextSpan, TextSpanIntervalIntrospector>? intervalTree)>();
+
+            public AnalyzerImpl(SimplifyTypeNamesDiagnosticAnalyzerBase<TLanguageKindEnum> analyzer)
+            {
+                _analyzer = analyzer;
+            }
+
+            public void AnalyzeCodeBlock(CodeBlockAnalysisContext context)
+            {
+                if (_analyzer.IsIgnoredCodeBlock(ref context))
+                    return;
+
+                var (completed, intervalTree) = _codeBlockIntervals.GetOrAdd(context.CodeBlock.SyntaxTree, _ => (new StrongBox<bool>(false), SimpleIntervalTree.Create(new TextSpanIntervalIntrospector(), Array.Empty<TextSpan>())));
+                if (completed.Value)
+                    return;
+
+                RoslynDebug.AssertNotNull(intervalTree);
+                lock (completed)
+                {
+                    if (completed.Value)
+                        return;
+
+                    if (intervalTree.HasIntervalThatOverlapsWith(context.CodeBlock.FullSpan.Start, context.CodeBlock.FullSpan.End))
+                        return;
+
+                    intervalTree.AddIntervalInPlace(context.CodeBlock.FullSpan);
+                }
+
+                _analyzer.AnalyzeCodeBlock(context);
+            }
+
+            public void AnalyzeSemanticModel(SemanticModelAnalysisContext context)
+            {
+                var (completed, intervalTree) = _codeBlockIntervals.GetOrAdd(context.SemanticModel.SyntaxTree, syntaxTree => (new StrongBox<bool>(true), null));
+                if (!completed.Value)
```

Comment:
wouldn't this always be false?  when would we have analyzesemanticmodel called twice with the same tree?  wouldn't that be a bug in the engine?  i'd prefer an assert/throw here.

Reply:
If `AnalyzeSemanticModel` gets called first (no prior code block callbacks), it creates the state with the initial value `true`.

Reply:
ah, interesting case.  can you doc that.  that was def subtle.

Reply:
Updated